include/spdk/env.h

```diff
@@ -323,6 +335,546 @@ int spdk_pci_addr_fmt(char *bdf, size_t sz, const struct spdk_pci_addr *addr);
+ *   The number of objects to add in the ring from the obj_table.
+ * \return
+ *   - n: Actual number of objects enqueued.
+ */
+unsigned
+spdk_ring_enqueue_burst(struct spdk_ring *sr, void *const *obj_table,
+			unsigned n);
+/**
+ * Dequeue several objects from a spdk ring (multi-consumers safe). When the request
+ * objects are more than the available objects, only dequeue the actual number
+ * of objects
+ *
+ * This function uses a "compare and set" instruction to move the
+ * consumer index atomically.
+ *
+ * \param sr
+ *   A pointer to the ring structure.
+ * \param obj_table
+ *   A pointer to a table of void * pointers (objects) that will be filled.
+ * \param n
+ *   The number of objects to dequeue from the ring to the obj_table.
+ * \return
+ *   - n: Actual number of objects dequeued, 0 if ring is empty
+ */
+unsigned
+spdk_ring_mc_dequeue_burst(struct spdk_ring *sr, void **obj_table, unsigned n);
+
+/**
+ * Dequeue several objects from a spdk ring (NOT multi-consumers safe).When the
+ * request objects are more than the available objects, only dequeue the
+ * actual number of objects
+ *
+ * \param sr
+ *   A pointer to the ring structure.
+ * \param obj_table
+ *   A pointer to a table of void * pointers (objects) that will be filled.
+ * \param n
+ *   The number of objects to dequeue from the ring to the obj_table.
+ * \return
+ *   - n: Actual number of objects dequeued, 0 if ring is empty
+ */
+unsigned
+spdk_ring_sc_dequeue_burst(struct spdk_ring *sr, void **obj_table, unsigned n);
+
+/**
+ * Dequeue multiple objects from a spdk ring up to a maximum number.
+ *
+ * This function calls the multi-consumers or the single-consumer
+ * version, depending on the default behaviour that was specified at
+ * ring creation time (see flags).
+ *
+ * \param sr
+ *   A pointer to the spdk ring structure.
+ * \param obj_table
+ *   A pointer to a table of void * pointers (objects) that will be filled.
+ * \param n
+ *   The number of objects to dequeue from the ring to the obj_table.
+ * \return
+ *   - Number of objects dequeued
+ */
+unsigned
+spdk_ring_dequeue_burst(struct spdk_ring *sr, void **obj_table, unsigned n);
+
+/**
+ * return the size of memory occupied by a ring
+ */
+ssize_t spdk_ring_get_memsize(unsigned count);
+
+/**
+ * Create a ring
+ */
+struct spdk_ring *
+spdk_ring_create(const char *name, unsigned count, int socket_id, unsigned flags);
+
+/**
+ * free the ring
+ */
+void spdk_ring_free(struct spdk_ring *sr);
+
+/**
+ * Change the high water mark. If *count* is 0, water marking is
+ * disabled
+ */
+int spdk_ring_set_water_mark(struct spdk_ring *sr, unsigned count);
+
+/**
+ * dump the status of the ring on the console
+ */
+void spdk_ring_dump(FILE *f, const struct spdk_ring *sr);
+
+/**
+ * dump the status of all rings on the console
+ */
+void spdk_ring_list_dump(FILE *f);
+
+/**
+ * search a ring from its name
+ */
+struct spdk_ring *spdk_ring_lookup(const char *name);
+
+#define SPDK_MAX_LCORE 128
```

Comment:
This is another DPDK configuration setting, so I don't think we want to hard-code 128 here.  It's already possible in extreme cases to have more than 128 logical cores on a single system (e.g. 8-socket systems).

This is a little trickier to avoid hard-coding, since we do want to use it to size static arrays; possibly we can replace those static arrays with dynamic allocations and turn this into an accessor function as well.